### PR TITLE
:sparkles: Image logging 

### PR DESCRIPTION
**Pull request by lalmei**

### CODE REVIEW

File: notebooks/Logging_Images.ipynb [hunk 1]
```diff
@@ -0,0 +1,478 @@
+{
```
Comment (andyndang) [inline notebooks/Logging_Images.ipynb:1]: FYI we kinda keep all the example notebooks under https://github.com/whylabs/whylogs-examples/tree/mainline/python at the moment.

Used to be mono-repo but we split things up for easier hacking things...

File: scripts/Untitled.ipynb [hunk 2]
```diff
@@ -0,0 +1,174 @@
+{
```
Comment (andyndang) [inline scripts/Untitled.ipynb]: Untitled?

File: src/whylogs/app/dataset.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        if self.root is not None:
+            body.append("Folder location: {}".format(self.root))
+        body += self.extra_repr().splitlines()
+        if hasattr(self, "transforms") and self.transforms is not None:
+            body += [repr(self.transforms)]
+        lines = [head] + [" " * self._repr_indent + line for line in body]
+        return '\n'.join(lines)
+
+
+class WhylogsDataset(Dataset):
```
Comment (andyndang) [inline src/whylogs/app/dataset.py]: we only use `whylogs` as lower case - it's an explicit decision :)

File: src/whylogs/app/dataset.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    def __init__(
+    self,
+    root_folder,
+    loader: Callable[[str],Any]= file_loader,
+    extensions: List[str] = EXTENSIONS,
+    transform:Optional[Callable] = None,
+    
+    is_valid_file: Optional[Callable[[str],bool]]= None,
+    ) -> None:
+    super(WhylogsDataset, obj).__init__(root_folder, transform= transform )
```
Comment (andyndang) [inline src/whylogs/app/dataset.py]: spacing here

File: src/whylogs/core/image_profiling.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+
+def image_loader(path: str = None)-> Image:
+
+    with open(path, "rb") as file_p:
+        img = Image.open(file_p).copy()
+        return img
+
+
+# def init():
```
Comment (andyndang) [inline src/whylogs/core/image_profiling.py]: remove redundant code?

File: src/whylogs/core/image_profiling.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    "Orientation",
+    "RowsPerStrip",
+    "ExposureTime",
+    "BrightnessValue",
+    "Flash",
+]
+
+
+def image_loader(path: str = None)-> Image:
+
```
Comment (andyndang) [inline src/whylogs/core/image_profiling.py]: nit: unnecessary blank line
```suggestion

```

File: src/whylogs/core/image_profiling.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        self.feature_transforms = feature_transforms
+
+        if feature_transforms is None:
+            self.feature_transforms = DEFAULT_IMAGE_FEATURES
+
+        self.feature_name = feature_name
+        self.metadata_attributes = metadata_attributes
+
+    def __call__(self, profiles):
+
```
Comment (andyndang) [inline src/whylogs/core/image_profiling.py]: blank line?

```suggestion
```

File: src/whylogs/core/image_profiling.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+# def init():
+
+#     for feature in _IMAGE_FEATURES:
+#         __import__(f"")
+
+class TrackImage:
+
+    def __init__(self,
```
Comment (andyndang) [inline src/whylogs/core/image_profiling.py:59]: can you add some documentation here?

File: src/whylogs/features/transforms.py [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+class Brightness:
+
+    """
+     Outputs the Brightness of each pixel in the image
+    """
+
+    def __call__(self, img: Union[Image, np.ndarray])->np.ndarray:
+        """
+        Args:
+            img (Union[Image, np.ndarray]): Description
```
Comment (andyndang) [inline src/whylogs/features/transforms.py]: meaningful documentation here?

File: src/whylogs/io/local_dataset.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
+import os
```
Comment (andyndang) [inline src/whylogs/io/local_dataset.py:1]: this file should be `local_image_dataset` - it's tightly coupled with PIL logic and imports

File: src/whylogs/io/local_dataset.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+EXTENSIONS = ('.csv', '.xls', '.jpg', '.jpeg', '.png', '.ppm', '.bmp',
+              '.pgm', '.tif', '.tiff', '.webp')
+
+
+def valid_file(fname):
+    extension = os.path.splitext(fname)[1]
+    return extension in EXTENSIONS
+
+
+def file_loader(path: str) -> Any:
```
Comment (andyndang) [inline src/whylogs/io/local_dataset.py:15]: this is not just a file loader -it's an image file loader, right?

File: src/whylogs/io/local_dataset.py [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+class Dataset(abc.ABC):
+
+    def __init__(self,
+                 root_folder: str = "",
+                 feature_transforms: Optional[List[Callable]] = None,
+                 )->None:
+        self.root_folder = os.path.expanduser(root_folder)
+        self.feature_transforms = feature_transforms
+
+    def __getitem__(self, index: int) -> Any:
```
Comment (andyndang) [inline src/whylogs/io/local_dataset.py:36]: consider using `@abc.abstractmethod` on top of this?

File: src/whylogs/io/local_dataset.py [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+        lines = [head] + [" " * self._repr_indent + line for line in body]
+        return '\n'.join(lines)
+
+
+class LocalDataset(Dataset):
+
+    def __init__(
+        self,
+        root_folder,
+        loader: Callable[[str], Any] = file_loader,
```
Comment (andyndang) [inline src/whylogs/io/local_dataset.py:60]: this implies heavy coupling with image logic?

File: src/whylogs/app/dataset.py [hunk 14]
```diff
@@ -0,0 +1,81 @@
+import os
+from typing import Any, Callable, List, Optional, Tuple
+import ABC
+
+from PIL import
```
Comment (andyndang) [inline src/whylogs/app/dataset.py]: This import creates an explicit dependency on PIL. Basically nobody can use this class unless they have PIL installed

File: src/whylogs/app/dataset.py [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+            sample = self.transform(sample)
+        if self.target_transform is not None:
+            target = self.target_transform(target)
+    def __len__(self)->int:
+        return len(self.data)
+
+
+def file_loader(path: str) -> Any:
+    # open path as file to avoid ResourceWarning (https://github.com/python-pillow/Pillow/issues/835)
+    check_extensions=
```
Comment (andyndang) [inline src/whylogs/app/dataset.py]: valid code?

File: src/whylogs/app/dataset.py (hunk 14 above)
Comment (andyndang) [inline src/whylogs/app/dataset.py]: also, valid import?

File: src/whylogs/core/image_profiling.py [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+                    each_profile.track(metadata)
+
+                else:
+                    for each_attr in self.metadata_attributes:
+                        attribute_value = metadata.get(each_attr, None)
+                        each_profile.track(
+                            self.feature_name+each_attr, attribute_value)
+
+
+def get_pil_image_metadata(img: Image)->Dict:
```
Comment (andyndang) [inline src/whylogs/core/image_profiling.py]: spacing

File: src/whylogs/features/transforms.py [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+
+class Hue:
+
+    def __call__(self, img: Union[Image, np.ndarray])->np.ndarray:
+        """
+        Args:
+            img (Union[Image, np.ndarray]): Description
+
+        Returns:
+            np.ndarray: Description
```
Comment (andyndang) [inline src/whylogs/features/transforms.py]: doc string

**Review by andyndang — CHANGES_REQUESTED** [review]

File: src/whylogs/app/dataset.py (hunk 4 above)
Comment (lalmei) [reply src/whylogs/app/dataset.py]:
```suggestion
    super(WhylogsDataset, obj).__init__(root_folder, transform = transform )
```

File: src/whylogs/app/dataset.py (hunk 3 above)
Comment (lalmei) [reply src/whylogs/app/dataset.py]: This is a class name, which by pep8 standards needs to be uppercase

Comment (lalmei) [reply src/whylogs/app/dataset.py]: will modify name

File: src/whylogs/core/image_profiling.py (hunk 16 above)
Comment (lalmei) [reply src/whylogs/core/image_profiling.py]:
```suggestion
def get_pil_image_metadata(img: Image) -> Dict:
```

File: notebooks/Logging_Images.ipynb (hunk 1 above)
Comment (lalmei) [reply notebooks/Logging_Images.ipynb:1]: Yeah the goal is to just speed it up, but also ideally we can keep track of notebook by running them during tests just to check if we are deprecating an example.

File: src/whylogs/features/transforms.py (hunk 17 above)
Comment (lalmei) [reply src/whylogs/features/transforms.py]: done

File: src/whylogs/app/logger.py [hunk 18]
```diff
@@ -348,6 +338,23 @@ def log_segment_datum(self, feature_name, value):
...
 
+    def log_image(self, image, feature_transforms: Optional[List[Callable]] = None, metadata_attributes: Optional[List[str]] = _METADATA_DEFAULT_ATTRIBUTES, feature_name: str = ""):
+
+        if not self._active:
+            return
+        if self.should_rotate():
+            self._rotate_time()
+
+        if isinstance(image, str):
+
```
Comment (andyndang) [inline src/whylogs/app/logger.py]:
```suggestion

```
Blank line

File: src/whylogs/features/transforms.py [hunk 19]
```diff
@@ -0,0 +1,125 @@
+
+import numpy as np
+from typing import Union, Callable, List
+from PIL.Image import Image as ImageType
```
Comment (andyndang) [inline src/whylogs/features/transforms.py:3]: FYI direct import from PIL will prevent users from accessing classes here

**Comment by andyndang** [comment]
FYI you'll want to add a section here with PIL as a dependency:
https://github.com/whylabs/whylogs-python/blob/mainline/setup.cfg#L87

**Review by andyndang — CHANGES_REQUESTED** [review]
Just feedback from documentation and direct import of PIL in some critical code paths.

I want to make sure we don't require PIL to run basic whylogs

File: src/whylogs/core/image_profiling.py (hunk 8 above)
Comment (lalmei) [reply src/whylogs/core/image_profiling.py:59]: done in e558a68d9c10385fa8853b8215ad94a03337557b

File: src/whylogs/features/transforms.py (hunk 9 above)
Comment (lalmei) [reply src/whylogs/features/transforms.py]: done

File: src/whylogs/features/transforms.py (hunk 19 above)
Comment (lalmei) [reply src/whylogs/features/transforms.py:3]: These are all image based, as we move to tensors we refactor it further.

File: src/whylogs/io/local_dataset.py (hunk 10 above)
Comment (lalmei) [reply src/whylogs/io/local_dataset.py:1]: For now, I will extend it to other file formats.

File: src/whylogs/io/local_dataset.py (hunk 11 above)
Comment (lalmei) [reply src/whylogs/io/local_dataset.py:15]: file loader, right now just image, will extend to other files.

File: src/whylogs/io/local_dataset.py (hunk 13 above)
Comment (lalmei) [reply src/whylogs/io/local_dataset.py:60]: any file, will start with images for now.

File: src/whylogs/io/local_dataset.py [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+
+class Dataset(abc.ABC):
+
+    def __init__(self,
+                 root_folder: str = "",
+                 feature_transforms: Optional[List[Callable]] = None,
+                 )->None:
+        self.root_folder = os.path.expanduser(root_folder)
+        self.feature_transforms = feature_transforms
+
```
Comment (lalmei) [inline src/whylogs/io/local_dataset.py]:
```suggestion
    @abc.abstractmethod
```

**Comment by andyndang** [comment]
I noticed `notebooks/Test.ipynb` is committed? Do we have a better name for the notebook?

**Review by andyndang — APPROVED** [review]
looking good

**Pull request merged**